components/range_cache_memory_engine/src/range_stats.rs

```diff
@@ -381,6 +337,98 @@ impl RangeStatsManager {
+            .info_provider
+            .get_regions_stat(cached_region_ids.clone())
+            .unwrap();
+        if regions_activity.is_empty() {
+            return;
+        }
+
+        // Use the average mvcc amplification to filter the regions with high mvcc
+        // amplification
+        let avg_mvcc_amplification = regions_activity
+            .iter()
+            .map(|(_, ra)| ra.cop_detail.mvcc_amplification())
+            .sum::<f64>()
+            / regions_activity.len() as f64;
+        let evict_candidates: Vec<_> = {
+            let mut regions_loaded = self.region_loaded_at.write().unwrap();
+            let mut evict_candidates = regions_activity;
+            evict_candidates.retain(|(r, ra)| {
+                let time_loaded = regions_loaded.entry(r.id).or_insert(Instant::now());
+                ra.cop_detail.mvcc_amplification() < avg_mvcc_amplification
+                    && Instant::now() - *time_loaded > self.evict_min_duration
+            });
+            evict_candidates.sort_by(|a, b| {
+                let next_prev_a = a.1.cop_detail.iterated_count();
+                let next_prev_b = b.1.cop_detail.iterated_count();
+                next_prev_a.cmp(&next_prev_b)
+            });
+
+            evict_candidates.into_iter().map(|(r, _)| r).collect()
+        };
+        // Evict two regions each time to reduce the probability that an un-dropped
+        // ongoing snapshot blocks the process
+        for regions in evict_candidates.chunks(2) {
+            let mut rxs = vec![];
+            let mut deleteable_regions = vec![];
+            for r in regions {
+                info!(
+                    "ime evict on soft limit reached";
+                    "region_to_evict" => ?r,
+                );
+
+                let (tx, rx) = sync_channel(1);
+                deleteable_regions.extend(evict_region(
+                    &CacheRegion::from_region(r),
+                    EvictReason::MemoryLimitReached,
+                    // This callback will be executed when eviction finishes at `on_delete_regions`
+                    // and when the reletive rx.recv() returns, we know some memory are freed.
+                    Some(Box::new(move || {
+                        let _ = tx.send(true);
+                    })),
+                ));
+                self.handle_region_evicted(r);
+                rxs.push(rx);
+            }
+            if !deleteable_regions.is_empty() {
+                if let Err(e) = delete_range_scheduler
+                    .schedule_force(BackgroundTask::DeleteRegions(deleteable_regions))
+                {
+                    error!(
+                        "ime schedule deletet range failed";
+                        "err" => ?e,
+                    );
+                    assert!(tikv_util::thread_group::is_shutdown(!cfg!(test)));
+                }
+            }
+            for rx in rxs {
+                let _ = rx.recv();
```

Comment:
Maybe better use `rx.recv_timeout()` to ensure in the worse case(maybe due to bug), it still not bug the thread forever.

Reply:
After changed to async callback, it does not support timeout now as it requires to use tokio runtime with timer enabled.